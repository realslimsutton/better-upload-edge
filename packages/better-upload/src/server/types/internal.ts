import type { ObjectCannedACL, StorageClass } from '@aws-sdk/client-s3';
import type { StandardSchemaV1 } from './standard-schema';

export type UnknownMetadata = Record<string, unknown>;
export type ObjectMetadata = Record<string, string>;

type ClientMetadata<T extends StandardSchemaV1 | undefined = undefined> =
  T extends StandardSchemaV1 ? StandardSchemaV1.InferOutput<T> : unknown;

export type FileInfo = {
  /**
   * The name of the file.
   */
  name: string;

  /**
   * File size in bytes.
   */
  size: number;

  /**
   * The MIME type of the file.
   */
  type: string;

  /**
   * The object key where the file will be uploaded to.
   */
  objectKey: string;
};

export type RouteConfig<
  Multiple extends boolean,
  Multipart extends boolean,
  InterMetadata extends UnknownMetadata,
  ClientMetadataSchema extends StandardSchemaV1 | undefined,
> = {
  /**
   * Maximum file size in bytes.
   *
   * @default 1024 * 1024 * 5 // 5MB
   *
   * @example
   *
   * ```ts
   * maxFileSize: 1024 * 1024 * 10 // 10MB
   * ```
   */
  maxFileSize?: number;

  /**
   * Allowed file types.
   *
   * @default [] // All file types are allowed
   *
   * @example
   *
   * ```ts
   * fileTypes: ['image/png', 'image/jpeg'] // Only PNG and JPEG files are allowed
   * ```
   *
   * @example
   *
   * ```ts
   * fileTypes: ['image/*'] // All image files are allowed
   * ```
   */
  fileTypes?: string[];

  /**
   * The number of seconds the upload signed URL is valid for.
   *
   * This is not used for multipart uploads.
   *
   * @default 120
   *
   * @example
   *
   * ```ts
   * signedUrlExpiresIn: 300 // 5 minutes
   * ```
   */
  signedUrlExpiresIn?: number;

  /**
   * Schema for validating metadata sent from the client. You can use any validation library that is compatible with `standard-schema`.
   *
   * @example
   *
   * ```ts
   * clientMetadataSchema: z.object({
   *  title: z.string(),
   * })
   * ```
   */
  clientMetadataSchema?: ClientMetadataSchema;

  /**
   * Use this callback to run custom logic before uploading a file, such as auth and rate-limiting. You can also return `objectInfo` to customize the S3 object (`generateObjectInfo` for multiple files). This runs only once regardless of the number of files uploaded.
   *
   * Metadata sent from the client is also available.
   *
   * Throw `RejectUpload` to reject the file upload. This will also send the error message to the client.
   */
  onBeforeUpload?: (
    data: {
      /**
       * The incoming request.
       */
      req: Request;

      /**
       * Metadata sent from the client.
       */
      clientMetadata: ClientMetadata<ClientMetadataSchema>;
    } & (Multiple extends false
      ? {
          /**
           * Information about the file to be uploaded.
           */
          file: Omit<FileInfo, 'objectKey'>;
        }
      : {
          /**
           * Information about the files to be uploaded.
           */
          files: Omit<FileInfo, 'objectKey'>[];
        })
  ) =>
    | BeforeUploadCallbackResult<Multiple, InterMetadata>
    | void
    | Promise<BeforeUploadCallbackResult<Multiple, InterMetadata> | void>;

  /**
   * Use this callback to run custom logic after creating the signed URL, such as saving data to a database. This runs only once regardless of the number of files uploaded.
   *
   * Metadata sent from `onBeforeUpload` is available as `metadata`. Metadata sent from the client is available as `clientMetadata`.
   *
   * You can return additional metadata to be sent back to the client, needs to be JSON serializable.
   */
  onAfterSignedUrl?: (
    data: {
      /**
       * The incoming request.
       */
      req: Request;

      /**
       * Metadata returned by `onBeforeUpload`.
       */
      metadata: InterMetadata;

      /**
       * Metadata sent from the client.
       */
      clientMetadata: ClientMetadata<ClientMetadataSchema>;
    } & (Multiple extends false
      ? {
          /**
           * Information about the uploaded file, including the object key.
           */
          file: FileInfo;
        }
      : {
          /**
           * Information about the uploaded files, including the object keys.
           */
          files: FileInfo[];
        })
  ) =>
    | AfterSignedUrlCallbackResult
    | void
    | Promise<AfterSignedUrlCallbackResult | void>;
} & (Multiple extends true
  ? {
      /**
       * Allow more than one file to be uploaded.
       *
       * @default false
       */
      multipleFiles: Multiple;

      /**
       * The maximum number of files that can be uploaded.
       *
       * @default 3
       */
      maxFiles?: number;
    }
  : {
      /**
       * Allow more than one file to be uploaded.
       *
       * @default false
       */
      multipleFiles?: Multiple;
    }) &
  (Multipart extends true
    ? {
        /**
         * Use multipart upload for large files.
         *
         * **Use this for files larger than 5GB.**
         *
         * @default false
         */
        multipart?: Multipart;

        /**
         * The size of each part in bytes.
         *
         * @default 1024 * 1024 * 50 // 50MB
         */
        partSize?: number;

        /**
         * The number of seconds the upload part pre-signed URL is valid for.
         *
         * @default 1500 // 25 minutes
         */
        partSignedUrlExpiresIn?: number;

        /**
         * The number of seconds the complete multipart upload pre-signed URL is valid for.
         *
         * @default 1800 // 30 minutes
         */
        completeSignedUrlExpiresIn?: number;
      }
    : {
        /**
         * Use multipart upload for large files.
         *
         * **Use this for files larger than 5GB.**
         *
         * @default false
         */
        multipart?: Multipart;
      });

export type BeforeUploadCallbackObjectInfo = {
  /**
   * The S3 object key to upload to.
   */
  key?: string;

  /**
   * Custom S3 object metadata.
   *
   * **All keys will be lower cased.**
   *
   * **WARNING:** All values here will be exposed to the client. Do not use this for sensitive data.
   */
  metadata?: ObjectMetadata;

  /**
   * ACL (access control list) to apply to the S3 object.
   */
  acl?: ObjectCannedACL;

  /**
   * Storage class to apply to the S3 object.
   */
  storageClass?: StorageClass;

  /**
   * Cache-Control header to apply to the S3 object.
   *
   * **WARNING:** If not set, the client is able to set it to any value (does not apply to multipart uploads).
   */
  cacheControl?: string;
};

type BeforeUploadCallbackResult<
  Multiple extends boolean,
  InterMetadata extends UnknownMetadata,
> = {
  /**
   * Metadata sent to `onAfterSignedUrl`.
   */
  metadata?: InterMetadata;

  /**
   * The bucket name to upload to.
   *
   * If you wish to upload to a different bucket than the one specified in the router.
   */
  bucketName?: string;
} & (Multiple extends false
  ? {
      /**
       * Use this to specify properties for the S3 object that will be uploaded.
       *
       * Options:
       * - `key`: The S3 object key to upload to.
       * - `metadata`: Custom S3 object metadata.
       * - `acl`: ACL to apply to the S3 object.
       * - `storageClass`: Storage class to apply to the S3 object.
       */
      objectInfo?: BeforeUploadCallbackObjectInfo;
    }
  : {
      /**
       * Use this to specify properties for the S3 objects that will be uploaded. Will be called for each file, in parallel.
       *
       * Options:
       * - `key`: The S3 object key to upload to.
       * - `metadata`: Custom S3 object metadata.
       * - `acl`: ACL to apply to the S3 object.
       * - `storageClass`: Storage class to apply to the S3 object.
       */
      generateObjectInfo?: (data: {
        /**
         * Information about the file to be uploaded.
         */
        file: Omit<FileInfo, 'objectKey'>;
      }) =>
        | BeforeUploadCallbackObjectInfo
        | Promise<BeforeUploadCallbackObjectInfo>;
    });

type AfterSignedUrlCallbackResult = {
  /**
   * Metadata sent back to the client.
   *
   * Needs to be JSON serializable.
   */
  metadata?: UnknownMetadata;
};

export type Route = {
  maxFileSize?: number;
  fileTypes?: string[];
  signedUrlExpiresIn?: number;
  clientMetadataSchema?: StandardSchemaV1;

  maxFiles?: number;

  multipart?: {
    partSize?: number;
    partSignedUrlExpiresIn?: number;
    completeSignedUrlExpiresIn?: number;
  };

  onBeforeUpload?: (data: {
    req: Request;
    clientMetadata: unknown;
    files: Omit<FileInfo, 'objectKey'>[];
  }) => Promise<{
    metadata?: UnknownMetadata;
    bucketName?: string;
    generateObjectInfo?: (data: {
      file: Omit<FileInfo, 'objectKey'>;
    }) =>
      | BeforeUploadCallbackObjectInfo
      | Promise<BeforeUploadCallbackObjectInfo>;
  } | void>;

  onAfterSignedUrl?: (data: {
    req: Request;
    metadata: UnknownMetadata;
    clientMetadata: unknown;
    files: FileInfo[];
  }) => Promise<{ metadata?: UnknownMetadata } | void>;
};

export type ExecRoute = () => Route;
