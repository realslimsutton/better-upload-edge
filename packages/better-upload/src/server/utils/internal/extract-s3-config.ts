/**
 * Extracts credentials and configuration from an S3Client.
 * Returns null if extraction fails (e.g., in Cloudflare Workers).
 */
export async function extractS3Config(client: any): Promise<{
  accessKeyId: string;
  secretAccessKey: string;
  region: string;
  endpoint: string;
} | null> {
  try {
    // Try to access the client's config
    // This works in Node.js environments
    const config = (client as any).config;
    if (!config) {
      return null;
    }

    // Get credentials
    const credentials = await config.credentials();
    if (!credentials?.accessKeyId || !credentials?.secretAccessKey) {
      return null;
    }

    // Get endpoint and region
    const endpoint =
      config.endpoint?.toString() ||
      config.endpointProvider?.({})?.url?.toString() ||
      `https://s3.${config.region || 'us-east-1'}.amazonaws.com`;

    const region = config.region || 'us-east-1';

    return {
      accessKeyId: credentials.accessKeyId,
      secretAccessKey: credentials.secretAccessKey,
      region,
      endpoint: endpoint.replace(/\/$/, ''),
    };
  } catch {
    // In Cloudflare Workers, S3Client won't work, so return null
    return null;
  }
}
