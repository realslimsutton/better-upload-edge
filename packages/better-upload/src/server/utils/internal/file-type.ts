export function isFileTypeAllowed(
  fileType: string,
  allowedFileTypes: string[]
) {
  let invalid = true;

  allowedFileTypes.forEach((type) => {
    if (type.endsWith('/*')) {
      const prefix = type.split('/*')[0];
      if (prefix && fileType.startsWith(prefix)) {
        invalid = false;
      }
    } else if (type === fileType) {
      invalid = false;
    }
  });

  return !invalid;
}
