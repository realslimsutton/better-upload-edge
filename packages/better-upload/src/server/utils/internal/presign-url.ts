/**
 * Creates a presigned URL for S3 using AWS Signature Version 4 query string signing.
 * Compatible with Cloudflare Workers.
 */

export interface PresignUrlParams {
  accessKeyId: string;
  secretAccessKey: string;
  region: string;
  endpoint: string;
  bucket: string;
  key: string;
  method?: string;
  expiresIn?: number;
  contentType?: string;
  contentLength?: number;
  metadata?: Record<string, string>;
  acl?: string;
  storageClass?: string;
  cacheControl?: string;
  signableHeaders?: string[];
}

/**
 * Creates a presigned URL for S3 operations using AWS Signature Version 4.
 */
export async function presignUrl(params: PresignUrlParams): Promise<string> {
  const {
    accessKeyId,
    secretAccessKey,
    region,
    endpoint,
    bucket,
    key,
    method = 'PUT',
    expiresIn = 120,
    contentType,
    contentLength,
    metadata = {},
    acl,
    storageClass,
    cacheControl,
    signableHeaders = [],
  } = params;

  // Normalize endpoint (remove trailing slash)
  const normalizedEndpoint = endpoint.replace(/\/$/, '');

  // Determine if using path-style or virtual-hosted-style URLs
  // Use path-style for: localhost, MinIO, Cloudflare R2, or buckets with dots
  const usePathStyle =
    normalizedEndpoint.includes('localhost') ||
    normalizedEndpoint.includes('127.0.0.1') ||
    normalizedEndpoint.includes('minio') ||
    normalizedEndpoint.includes('r2.cloudflarestorage.com') ||
    bucket.includes('.');

  const host = usePathStyle
    ? normalizedEndpoint.replace(/^https?:\/\//, '')
    : `${bucket}.${normalizedEndpoint.replace(/^https?:\/\//, '')}`;

  const path = usePathStyle ? `/${bucket}/${key}` : `/${key}`;

  // Build query parameters
  const now = new Date();
  const dateStamp = now.toISOString().slice(0, 10).replace(/-/g, '');
  const amzDate = `${dateStamp}T${now.toISOString().slice(11, 19).replace(/:/g, '')}Z`;
  const credentialScope = `${dateStamp}/${region}/s3/aws4_request`;

  const queryParams: Record<string, string> = {
    'X-Amz-Algorithm': 'AWS4-HMAC-SHA256',
    'X-Amz-Credential': `${accessKeyId}/${credentialScope}`,
    'X-Amz-Date': amzDate,
    'X-Amz-Expires': expiresIn.toString(),
  };

  // Add metadata headers
  Object.entries(metadata).forEach(([k, v]) => {
    const headerKey = `x-amz-meta-${k.toLowerCase()}`;
    queryParams[headerKey] = v;
  });

  // Build headers that will be signed
  const headers: Record<string, string> = {
    host,
  };

  if (contentType) {
    headers['content-type'] = contentType;
  }

  if (contentLength !== undefined) {
    headers['content-length'] = contentLength.toString();
  }

  if (cacheControl && signableHeaders.includes('cache-control')) {
    headers['cache-control'] = cacheControl;
    queryParams['Cache-Control'] = cacheControl;
  }

  if (acl) {
    queryParams['x-amz-acl'] = acl;
  }

  if (storageClass) {
    queryParams['x-amz-storage-class'] = storageClass;
  }

  // Build canonical request
  const canonicalHeaders = Object.keys(headers)
    .sort()
    .map((k) => `${k.toLowerCase()}:${headers[k]}`)
    .join('\n');
  const signedHeaders = Object.keys(headers)
    .sort()
    .map((k) => k.toLowerCase())
    .join(';');

  // Build query string (sorted)
  const queryString = Object.keys(queryParams)
    .sort()
    .map(
      (k) => `${encodeURIComponent(k)}=${encodeURIComponent(queryParams[k])}`
    )
    .join('&');

  const canonicalRequest = [
    method,
    path,
    queryString,
    canonicalHeaders,
    '',
    signedHeaders,
    'UNSIGNED-PAYLOAD',
  ].join('\n');

  // Create string to sign
  const stringToSign = [
    'AWS4-HMAC-SHA256',
    amzDate,
    credentialScope,
    await sha256(canonicalRequest),
  ].join('\n');

  // Calculate signature
  const kDate = await hmacSha256Raw(`AWS4${secretAccessKey}`, dateStamp);
  const kRegion = await hmacSha256Raw(kDate, region);
  const kService = await hmacSha256Raw(kRegion, 's3');
  const kSigning = await hmacSha256Raw(kService, 'aws4_request');
  const signature = await hmacSha256Hex(kSigning, stringToSign);

  // Add signature to query params
  queryParams['X-Amz-Signature'] = signature;

  // Build final URL
  const finalQueryString = Object.keys(queryParams)
    .sort()
    .map(
      (k) => `${encodeURIComponent(k)}=${encodeURIComponent(queryParams[k])}`
    )
    .join('&');

  const protocol = normalizedEndpoint.startsWith('https') ? 'https' : 'http';
  return `${protocol}://${host}${path}?${finalQueryString}`;
}

// Helper functions for cryptographic operations using Web Crypto API
async function sha256(message: string): Promise<string> {
  const encoder = new TextEncoder();
  const data = encoder.encode(message);
  const hashBuffer = await crypto.subtle.digest('SHA-256', data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');
}

async function hmacSha256Raw(
  key: string | ArrayBuffer,
  message: string
): Promise<ArrayBuffer> {
  const encoder = new TextEncoder();
  let keyData: CryptoKey;

  if (typeof key === 'string') {
    const keyBuffer = encoder.encode(key);
    keyData = await crypto.subtle.importKey(
      'raw',
      keyBuffer,
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign']
    );
  } else {
    keyData = await crypto.subtle.importKey(
      'raw',
      key,
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign']
    );
  }

  const messageData = encoder.encode(message);
  return await crypto.subtle.sign('HMAC', keyData, messageData);
}

async function hmacSha256Hex(
  key: ArrayBuffer,
  message: string
): Promise<string> {
  const encoder = new TextEncoder();
  const keyData = await crypto.subtle.importKey(
    'raw',
    key,
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign']
  );

  const messageData = encoder.encode(message);
  const signatureBuffer = await crypto.subtle.sign(
    'HMAC',
    keyData,
    messageData
  );
  return Array.from(new Uint8Array(signatureBuffer))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}
