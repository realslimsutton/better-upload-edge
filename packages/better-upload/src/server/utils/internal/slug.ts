export function createSlug(text: string) {
  return text
    .normalize('NFD') // split an accented letter in the base letter and the acent
    .replace(/[\u0300-\u036f]/g, '') // remove all previously split accents
    .toLowerCase()
    .trim()
    .replace(/-/g, ' ') // replace dashes with spaces
    .replace(/[^a-z0-9. ]/g, '') // remove all chars not letters, numbers and spaces (to be replaced)
    .replace(/\s+/g, '-');
}
