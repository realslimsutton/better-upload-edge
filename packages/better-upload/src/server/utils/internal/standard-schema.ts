import type { StandardSchemaV1 } from '@/server/types/standard-schema';

export async function standardValidate<T extends StandardSchemaV1>(
  schema: T,
  input: unknown
): Promise<StandardSchemaV1.Result<StandardSchemaV1.InferOutput<T>>> {
  let result = schema['~standard'].validate(input);
  if (result instanceof Promise) result = await result;

  return result;
}
