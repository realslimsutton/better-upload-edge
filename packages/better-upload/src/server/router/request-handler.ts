import type { S3Client } from '@aws-sdk/client-s3';
import type { Router, S3Credentials } from '../types/public';
import { extractS3Config } from '../utils/internal/extract-s3-config';
import { standardValidate } from '../utils/internal/standard-schema';
import { uploadFileSchema } from '../validations';
import { handleFiles } from './handlers/files-handler';

export async function handleRequest(req: Request, router: Router) {
  if (req.method !== 'POST') {
    return Response.json(
      {
        error: {
          type: 'invalid_request',
          message: 'Method not allowed.',
        },
      },
      { status: 405 }
    );
  }

  let body;
  try {
    body = await req.json();
  } catch (error) {
    return Response.json(
      {
        error: {
          type: 'invalid_request',
          message: 'Invalid JSON body.',
        },
      },
      { status: 400 }
    );
  }

  const parsed = uploadFileSchema.safeParse(body);
  if (!parsed.success) {
    return Response.json(
      {
        error: {
          type: 'invalid_request',
          message: 'Invalid file upload schema.',
        },
      },
      { status: 400 }
    );
  }

  if (!(parsed.data.route in router.routes)) {
    return Response.json(
      {
        error: {
          type: 'invalid_request',
          message: 'Upload route not found.',
        },
      },
      { status: 404 }
    );
  }

  const route = router.routes[parsed.data.route]!();

  if (route.maxFiles === 1 && parsed.data.files.length > 1) {
    return Response.json(
      {
        error: {
          type: 'too_many_files',
          message: 'Multiple files are not allowed.',
        },
      },
      { status: 400 }
    );
  }

  let clientMetadata = parsed.data.metadata;
  if (route.clientMetadataSchema) {
    const validation = await standardValidate(
      route.clientMetadataSchema,
      clientMetadata
    );

    if (validation.issues) {
      return Response.json(
        {
          error: {
            type: 'invalid_request',
            message: 'Invalid metadata.',
          },
        },
        { status: 400 }
      );
    }
  }

  const data = {
    ...parsed.data,
    metadata: clientMetadata,
  };

  // Extract credentials - handle both S3Client and direct credentials
  let credentials: S3Credentials | undefined;
  let client: S3Client | undefined;

  if ('accessKeyId' in router.client) {
    // Direct credentials provided (for Cloudflare Workers)
    credentials = router.client as S3Credentials;
  } else {
    // S3Client provided (for Node.js environments)
    client = router.client as S3Client;
    credentials = (await extractS3Config(client)) || undefined;
  }

  // if (route.multipart) {
  //   return handleMultipartFiles({
  //     req,
  //     client: client,
  //     defaultBucketName: router.bucketName,
  //     route,
  //     data,
  //   });
  // }

  return handleFiles({
    req,
    client: client,
    credentials: credentials,
    defaultBucketName: router.bucketName,
    route,
    data,
  });
}
