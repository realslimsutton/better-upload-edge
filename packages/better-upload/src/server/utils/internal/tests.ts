import { RejectUpload } from '@/server/error';
import type { Router } from '@/server/types/public';
import type { UploadFileSchema } from '@/server/validations';
import { S3Client } from '@aws-sdk/client-s3';
import { z } from 'zod/mini';
import { route } from '../router';

export const testRouter: Router = {
  client: new S3Client({
    region: 'us-east-1',
    credentials: {
      accessKeyId: 'test',
      secretAccessKey: 'test',
    },
  }),
  bucketName: 'example-bucket',
  routes: {
    multipleImages: route({
      multipleFiles: true,
      maxFiles: 3,
      maxFileSize: 1024 * 1024 * 5, // 5MB
      fileTypes: ['image/*'],
      onBeforeUpload() {
        return {
          generateObjectInfo: ({ file }) => ({ key: `multiple/${file.name}` }),
        };
      },
    }),
    singleImage: route({
      maxFileSize: 1024 * 1024 * 5, // 5MB
      fileTypes: ['image/*'],
      onBeforeUpload({ file }) {
        return {
          objectInfo: { key: `single/${file.name}` },
        };
      },
    }),
    withMetaSchema: route({
      clientMetadataSchema: z.object({
        name: z.string(),
      }),
    }),
    alwaysReject: route({
      onBeforeUpload() {
        throw new RejectUpload('Test reject');
      },
    }),
    customBucket: route({
      onBeforeUpload({ file }) {
        return {
          bucketName: 'my-custom-bucket',
          objectInfo: { key: `custom/${file.name}` },
        };
      },
    }),
  },
};

export const testRequest = (opts: RequestInit) =>
  new Request('http://localhost:3000/api/upload', opts);

export const uploadBody = (body: UploadFileSchema) => JSON.stringify(body);
