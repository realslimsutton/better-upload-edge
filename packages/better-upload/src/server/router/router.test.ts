import { describe, expect, it } from 'vitest';
import { testRequest, testRouter, uploadBody } from '../utils/internal/tests';
import { handleRequest } from './request-handler';

describe('router', () => {
  it('should reject invalid method', async () => {
    const res = await handleRequest(testRequest({ method: 'GET' }), testRouter);
    const json = await res.json();

    expect(res.status).toBe(405);
    expect(json).toEqual({
      error: {
        type: 'invalid_request',
        message: 'Method not allowed.',
      },
    });
  });

  it('should reject invalid JSON body', async () => {
    const res = await handleRequest(
      testRequest({ method: 'POST', body: 'invalid-json' }),
      testRouter
    );
    const json = await res.json();

    expect(res.status).toBe(400);
    expect(json).toEqual({
      error: {
        type: 'invalid_request',
        message: 'Invalid JSON body.',
      },
    });
  });

  it('should reject invalid upload schema', async () => {
    const res = await handleRequest(
      testRequest({
        method: 'POST',
        body: JSON.stringify({ invalid: 'data' }),
      }),
      testRouter
    );
    const json = await res.json();

    expect(res.status).toBe(400);
    expect(json).toEqual({
      error: {
        type: 'invalid_request',
        message: 'Invalid file upload schema.',
      },
    });
  });

  it('should reject non-existing upload route', async () => {
    const res = await handleRequest(
      testRequest({
        method: 'POST',
        body: uploadBody({
          route: 'nonExisting',
          files: [{ name: 'file1.jpg', size: 500000, type: 'image/jpeg' }],
        }),
      }),
      testRouter
    );
    const json = await res.json();

    expect(res.status).toBe(404);
    expect(json).toEqual({
      error: {
        type: 'invalid_request',
        message: 'Upload route not found.',
      },
    });
  });

  it('should reject multiple files on single file route', async () => {
    const res = await handleRequest(
      testRequest({
        method: 'POST',
        body: uploadBody({
          route: 'singleImage',
          files: [
            { name: 'file1.jpg', size: 500000, type: 'image/jpeg' },
            { name: 'file2.jpg', size: 400000, type: 'image/jpeg' },
          ],
        }),
      }),
      testRouter
    );
    const json = await res.json();

    expect(res.status).toBe(400);
    expect(json).toEqual({
      error: {
        type: 'too_many_files',
        message: 'Multiple files are not allowed.',
      },
    });
  });

  it('should reject invalid client metadata', async () => {
    const res = await handleRequest(
      testRequest({
        method: 'POST',
        body: uploadBody({
          route: 'withMetaSchema',
          files: [{ name: 'file1.jpg', size: 500000, type: 'image/jpeg' }],
          metadata: { name: 123 }, // Invalid, should be string
        }),
      }),
      testRouter
    );
    const json = await res.json();

    expect(res.status).toBe(400);
    expect(json).toEqual({
      error: {
        type: 'invalid_request',
        message: 'Invalid metadata.',
      },
    });
  });

  describe('non multipart handler', () => {
    it('should reject too many files', async () => {
      const res = await handleRequest(
        testRequest({
          method: 'POST',
          body: uploadBody({
            route: 'multipleImages',
            files: [
              { name: 'file1.jpg', size: 500000, type: 'image/jpeg' },
              { name: 'file2.jpg', size: 400000, type: 'image/jpeg' },
              { name: 'file3.jpg', size: 300000, type: 'image/jpeg' },
              { name: 'file4.jpg', size: 200000, type: 'image/jpeg' },
            ],
          }),
        }),
        testRouter
      );
      const json = await res.json();

      expect(res.status).toBe(400);
      expect(json).toEqual({
        error: {
          type: 'too_many_files',
          message: 'Too many files.',
        },
      });
    });

    it('should reject file too large', async () => {
      const res = await handleRequest(
        testRequest({
          method: 'POST',
          body: uploadBody({
            route: 'singleImage',
            files: [
              { name: 'file1.jpg', size: 1024 * 1024 * 10, type: 'image/jpeg' },
            ],
          }),
        }),
        testRouter
      );
      const json = await res.json();

      expect(res.status).toBe(400);
      expect(json).toEqual({
        error: {
          type: 'file_too_large',
          message: 'One or more files are too large.',
        },
      });
    });

    it('should reject S3 file size 5gb limit', async () => {
      const res = await handleRequest(
        testRequest({
          method: 'POST',
          body: uploadBody({
            route: 'singleImage',
            files: [
              {
                name: 'file1.jpg',
                size: 1024 * 1024 * 6000,
                type: 'image/jpeg',
              },
            ],
          }),
        }),
        testRouter
      );
      const json = await res.json();

      expect(res.status).toBe(400);
      expect(json).toEqual({
        error: {
          type: 'file_too_large',
          message:
            'One or more files exceed the S3 limit of 5GB. Use multipart upload for larger files.',
        },
      });
    });

    it('should reject invalid file type', async () => {
      const res = await handleRequest(
        testRequest({
          method: 'POST',
          body: uploadBody({
            route: 'singleImage',
            files: [
              { name: 'file1.pdf', size: 500000, type: 'application/pdf' },
            ],
          }),
        }),
        testRouter
      );
      const json = await res.json();

      expect(res.status).toBe(400);
      expect(json).toEqual({
        error: {
          type: 'invalid_file_type',
          message: 'One or more files have an invalid file type.',
        },
      });
    });

    it('should reject upload in onBeforeUpload', async () => {
      const res = await handleRequest(
        testRequest({
          method: 'POST',
          body: uploadBody({
            route: 'alwaysReject',
            files: [{ name: 'file1.jpg', size: 500000, type: 'image/jpeg' }],
          }),
        }),
        testRouter
      );
      const json = await res.json();

      expect(res.status).toBe(400);
      expect(json).toEqual({
        error: {
          type: 'rejected',
          message: 'Test reject',
        },
      });
    });

    it('should accept valid upload with custom bucket', async () => {
      const res = await handleRequest(
        testRequest({
          method: 'POST',
          body: uploadBody({
            route: 'customBucket',
            files: [{ name: 'file1.jpg', size: 500000, type: 'image/jpeg' }],
          }),
        }),
        testRouter
      );
      const json = await res.json();

      expect(json.files[0].signedUrl).toContain('my-custom-bucket');

      json.files[0].signedUrl = 'signed-url-placeholder';

      expect(res.status).toBe(200);
      expect(json).toMatchSnapshot();
    });

    it('should accept valid single file upload request', async () => {
      const res = await handleRequest(
        testRequest({
          method: 'POST',
          body: uploadBody({
            route: 'singleImage',
            files: [{ name: 'file1.jpg', size: 500000, type: 'image/jpeg' }],
          }),
        }),
        testRouter
      );
      const json = await res.json();

      json.files[0].signedUrl = 'signed-url-placeholder';

      expect(res.status).toBe(200);
      expect(json).toMatchSnapshot();
    });

    it('should accept valid multiple files upload request', async () => {
      const res = await handleRequest(
        testRequest({
          method: 'POST',
          body: uploadBody({
            route: 'multipleImages',
            files: [
              { name: 'file1.jpg', size: 500000, type: 'image/jpeg' },
              { name: 'file2.png', size: 400000, type: 'image/png' },
            ],
          }),
        }),
        testRouter
      );
      const json = await res.json();

      json.files[0].signedUrl = 'signed-url-placeholder';
      json.files[1].signedUrl = 'signed-url-placeholder';

      expect(res.status).toBe(200);
      expect(json).toMatchSnapshot();
    });
  });

  // describe('multipart handler', () => {
  //   // Multipart handler tests go here
  // });
});
