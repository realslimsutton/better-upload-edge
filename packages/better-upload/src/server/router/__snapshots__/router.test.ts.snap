// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`router > non multipart handler > should accept valid multiple files upload request 1`] = `
{
  "files": [
    {
      "file": {
        "name": "file1.jpg",
        "objectKey": "multiple/file1.jpg",
        "objectMetadata": {},
        "size": 500000,
        "type": "image/jpeg",
      },
      "signedUrl": "signed-url-placeholder",
    },
    {
      "file": {
        "name": "file2.png",
        "objectKey": "multiple/file2.png",
        "objectMetadata": {},
        "size": 400000,
        "type": "image/png",
      },
      "signedUrl": "signed-url-placeholder",
    },
  ],
  "metadata": {},
}
`;

exports[`router > non multipart handler > should accept valid single file upload request 1`] = `
{
  "files": [
    {
      "file": {
        "name": "file1.jpg",
        "objectKey": "single/file1.jpg",
        "objectMetadata": {},
        "size": 500000,
        "type": "image/jpeg",
      },
      "signedUrl": "signed-url-placeholder",
    },
  ],
  "metadata": {},
}
`;

exports[`router > non multipart handler > should accept valid upload with custom bucket 1`] = `
{
  "files": [
    {
      "file": {
        "name": "file1.jpg",
        "objectKey": "custom/file1.jpg",
        "objectMetadata": {},
        "size": 500000,
        "type": "image/jpeg",
      },
      "signedUrl": "signed-url-placeholder",
    },
  ],
  "metadata": {},
}
`;
